File: kotlin/lesson 3/SubarraySum.kt
class Solution {
    fun subarraySum(nums: IntArray, k: Int): Int {
        val sums = mutableMapOf(0 to 1)        
        var result = 0
        var sum = 0
        for (num in nums) {            
            sum += num
            result += sums.getOrDefault(sum - k, 0)
            sums[sum] = sums.getOrDefault(sum, 0) + 1
        }
        
        return result
    }
}

File: ruby/lesson 3/subarray_sum.rb
# https://leetcode.com/problems/subarray-sum-equals-k/
def subarray_sum(nums, k)
    sums = { 0 => 1 }
    result = sum = 0
    nums.each do |num|
        sum += num
        result += sums.fetch(sum - k, 0)
        sums[sum] = sums.fetch(sum, 0) + 1
    end
    result
end
